code/modules/hydroponics/biogenerator.dm

```diff
@@ -1,337 +1,354 @@
+		to_chat(user, "<span class='warning'>You cannot put this in [name]!</span>")
 
-	var/datum/browser/popup = new(user, "biogen", name, 350, 520)
-	popup.set_content(dat)
-	popup.open()
-	return
+/**
+ * Builds/Updates the `product_list` used by the UI.
+ */
+/obj/machinery/biogenerator/proc/update_ui_product_list()
+	product_list = list()
+	for(var/category in categories)
+		product_list[category] = list()
+
+	for(var/V in files.known_designs)
+		var/datum/design/D = files.known_designs[V]
+		for(var/category in categories)
+			if(!(category in D.category))
+				continue
+			product_list[category][D.name] = list(
+				"name" = D.name,
+				"id" = D.id,
+				"cost" = D.materials[MAT_BIOMASS]
+			)
 
 /obj/machinery/biogenerator/attack_hand(mob/user)
-	interact(user)
+	ui_interact(user)
 
 /obj/machinery/biogenerator/attack_ghost(mob/user)
-	interact(user)
+	ui_interact(user)
+
+/obj/machinery/biogenerator/ui_interact(mob/user, ui_key = "main", datum/tgui/ui = null, force_open = FALSE, datum/tgui/master_ui = null, datum/ui_state/state = GLOB.default_state)
+	ui = SStgui.try_update_ui(user, src, ui_key, ui, force_open)
+	if(!ui)
+		ui = new(user, src, ui_key, "Biogenerator", "Biogenerator", 390, 600, master_ui, state)
+		ui.set_autoupdate(FALSE)
+		ui.open()
+
+/obj/machinery/biogenerator/ui_data(mob/user)
+	var/list/data = list(
+		"processing" = processing,
+		"biomass" = biomass,
+		"efficiency" = efficiency,
+		"has_plants" = length(stored_plants) ? TRUE : FALSE
+	)
+	if(container)
+		data["container"] = TRUE
+		data["container_curr_reagents"] = container.reagents.total_volume
+		data["container_max_reagents"] = container.reagents.maximum_volume
+	else
+		data["container"] = FALSE
+
+	return data
 
-/obj/machinery/biogenerator/proc/activate()
-	if(usr.stat != 0)
+/obj/machinery/biogenerator/ui_static_data(mob/user)
+	return list("product_list" = product_list)
+
+/obj/machinery/biogenerator/ui_act(action, list/params)
+	if(..())
 		return
-	if(stat != 0) //NOPOWER etc
+
+	. = TRUE
+	switch(action)
+		if("activate")
+			activate(usr)
+		if("detach_container")
+			detach_container()
+		if("eject_plants")
+			eject_plants()
+		if("create")
+			var/datum/design/D = files.known_designs[params["id"]]
+			if(!D)
+				return
+			var/amount = clamp(text2num(params["amount"]), 1, 10)
+			create_product(D, amount)
+
+/**
+ * Tells the biogenerator to grind up any stored plants, and produce an appropriate amount of biomass.
+ *
+ * Argumens:
+ * * mob/user - the mob who activated the biogenerator
+ */
+/obj/machinery/biogenerator/proc/activate(mob/user)
+	if(stat & (NOPOWER | BROKEN))
 		return
 	if(processing)
-		to_chat(usr, "<span class='warning'>The biogenerator is in the process of working.</span>")
+		to_chat(user, "<span class='warning'>[src] is currently processing!</span>")
 		return
+
+	processing = TRUE
+	SStgui.update_uis(src)
+	update_icon()
+
 	var/S = 0
```

Comment:
make it `var/plants_processed = length(stored_plants)`, andjust the `use_power` and `sleep` accordingly.
This current math is way more complicated than it has any right to be